src/misc/unicode.cpp

```diff
@@ -0,0 +1,1129 @@
+/*
+ *  Copyright (C) 2022       The DOSBox Staging Team
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "dosbox.h"
+
+#include <algorithm>
+#include <fstream>
+#include <map>
+#include <regex>
+#include <set>
+#include <string>
+#include <vector>
+
+#include "checks.h"
+#include "dos_inc.h"
+#include "string_utils.h"
+
+CHECK_NARROWING();
+
+
+class Grapheme final {
+
+public:
+
+    Grapheme() = default;
+    Grapheme(const uint16_t code_point);
+
+    bool IsEmpty() const;
+    bool IsValid() const;
+    bool HasMark() const;
+    uint16_t GetCodePoint() const;
+
+    void Invalidate();
+    void AddMark(const uint16_t code_point);
+    void StripMarks();
+
+    bool operator<(const Grapheme &other) const;
+
+private:
+
+    // Unicode code point
+    uint16_t code_point = static_cast<uint16_t>(' ');
```

Comment:
Very minor thing, but just wondering why you're defaulting to a space character?

Reply:
That's because the default constructor creates an empty/undefined grapheme, and (at least with FreeDOS CPX files) undefined characters are actually spaces - here is an example ASCII table printed when code page is set to 869, which contains several undefined characters: 

![CP869](https://user-images.githubusercontent.com/48332137/187483844-2f8e0f99-70b8-44da-83a9-020ac6221123.png)